Remove global 'using namespace std;' definitions

Alhetus opened the pull request:
Using `using namespace std;` clauses in header files will pollute the whole program with all the std namespace functions, and will also bleed into third party libraries. This could cause ambiguity problems with some libraries in the future or the program code in general.

https://stackoverflow.com/questions/1452721/why-is-using-namespace-std-considered-bad-practice

Alhetus committed:
Ignore JetBrains' IDE files in .gitignore

Alhetus committed:
Fix warning generated by FindIconv.cmake

teejusb commented:
Heck yeaaaaah. I sort of peeked at the ITGmania forks and saw this earlier :)

Gonna let the integration checks run but I think this is good to me! I assume the methodology was to removing `using namespace std;` in `global.h` and then prefixed std functions/types with `std::`. The review would be kind of hard to do for 500+ files so wanted to make sure.

Alhetus commented:
This will probably need some testing as well on all platforms. All the builds pass but I might’ve done something stupid by mistake.

But yeah essentially what was done here:
- removed all the `using namespace std;` definitions, I think there were 3 in total.
- Removed some inline global min and max function overloads so now everything needs to be strictly typed and use purely std library functions.
- Prefixed all the std function calls with `std::`. Also some whitespace changes and type definitions/casts for min & max function calls as required.
- Got rid of the annoying Iconv cmake warning.
- Added .idea folder to .gitignore to support JetBrains IDE’s

natano commented:
I fully agree with removing using namespace std;. I had this on my mental todo list for a while, just no time to actually do it. Two comments about other things though:
- Im not sure the iconv change is correct. I would have to look more into it, but if you are confident it's correct feel free to merge. @teejusb 
- there seems to be weird whitespace changes in the diff when looking at it in GH. Maybe that's tabs vs. spaces or something? (ITGm uses tabs)

Alhetus commented:
> I fully agree with removing using namespace std;. I had this on my mental todo list for a while, just no time to actually do it. Two comments about other things though:
> 
> * Im not sure the iconv change is correct. I would have to look more into it, but if you are confident it's correct feel free to merge. @teejusb
> * there seems to be weird whitespace changes in the diff when looking at it in GH. Maybe that's tabs vs. spaces or something? (ITGm uses tabs)

Yeah, I probably have some spaces in there instead of tabs as that's the default in my IDE. I've noticed there are multiple different conventions in the codebase, which is to be expected due to SM's history. But it would be nice to do a complete code style guide, do config files for most IDE's and then roll a style fix for the whole codebase in one commit to get it all sorted.

natano commented:
I looked a bit more into FindIconv.cmake and I think your fix is correct. FTR, my concern was regarding the naming of the output variables.

Regarding whitespace: Where have you seen different conventions in the codebase? Every cpp file in the SM codebase I've opened so far uses tabs for indentation. I agree with fixing the formatting in the future, but as we don't have any automatic formatting yet it should be fixed for this PR. (For clarification: the indentation style should match the rest of the file. If you _do_ find a file that uses whitespace for indentation use whitespace, but otherwise tabs.)

Alhetus commented:
> I looked a bit more into FindIconv.cmake and I think your fix is correct. FTR, my concern was regarding the naming of the output variables.
> 
> Regarding whitespace: Where have you seen different conventions in the codebase? Every cpp file in the SM codebase I've opened so far uses tabs for indentation. I agree with fixing the formatting in the future, but as we don't have any automatic formatting yet it should be fixed for this PR. (For clarification: the indentation style should match the rest of the file. If you _do_ find a file that uses whitespace for indentation use whitespace, but otherwise tabs.)

If you do a `grep -inr --include \*.h --include \*.cpp '^    ' ./` (<- there should be 4 spaces after the ^ but github auto-formatting bleh...) in the src directory, there are a lot of files that have some lines that are indented with spaces instead of tabs. Of course files that are fully indented with spaces are very rare. With differing conventions I meant more of the style of function definitions and calls. For example, there are a lot of cases where `foo( int bar, int baz );` are used and a lot of cases where `foo(int bar, int baz);` are used. And the same for pointer markings, some files use `int *foo` and some `int* foo`. But I'll try to go through all the changes in this PR and revert the spaces to tabs.

Alhetus committed:
Remove global "using namespace std;" declarations, use "std::" prefixes on all std elements


Fix whitespace changes

Alhetus commented:
All the tabs should be restored now.

natano commented:
Looks good now, thanks for the PR @Alhetus!
